pruner/checkpoint.go

```diff
@@ -23,25 +30,60 @@ type checkpoint struct {
 	FailedHeaders    map[uint64]struct{} `json:"failed"`
 }
 
-// initializeCheckpoint initializes the checkpoint, storing the earliest header in the chain.
-func (s *Service) initializeCheckpoint(ctx context.Context) error {
-	return s.updateCheckpoint(ctx, uint64(1), nil)
+// DetectPreviousRun checks if the pruner has run before by checking for the existence of a
+// checkpoint.
+func DetectPreviousRun(ctx context.Context, ds datastore.Datastore) error {
+	_, err := getCheckpoint(ctx, namespace.Wrap(ds, storePrefix))
+	if errors.Is(err, errCheckpointNotFound) {
+		return nil
+	}
+	if err != nil {
+		return fmt.Errorf("failed to load checkpoint: %w", err)
+	}
+	return ErrDisallowRevertToArchival
 }
 
-// loadCheckpoint loads the last checkpoint from disk, initializing it if it does not already exist.
-func (s *Service) loadCheckpoint(ctx context.Context) error {
-	bin, err := s.ds.Get(ctx, checkpointKey)
+// storeCheckpoint persists the checkpoint to disk.
+func storeCheckpoint(ctx context.Context, ds datastore.Datastore, c *checkpoint) error {
+	bin, err := json.Marshal(c)
+	if err != nil {
+		return err
+	}
+
+	return ds.Put(ctx, checkpointKey, bin)
+}
+
+// getCheckpoint loads the last checkpoint from disk.
+func getCheckpoint(ctx context.Context, ds datastore.Datastore) (*checkpoint, error) {
+	bin, err := ds.Get(ctx, checkpointKey)
 	if err != nil {
 		if errors.Is(err, datastore.ErrNotFound) {
-			return s.initializeCheckpoint(ctx)
+			return nil, errCheckpointNotFound
```

Comment:
why not just check for datastore.ErrNotFound err in `DetectPreviousRun` ?

Reply:
the error is also used for init of checkpoint. I think it is easier to read via err name this way:
```
if err.Is(errCheckpointNotFound){
  initCheckpoint
}
```
instead of 
```
if err.Is(dagstpre.ErrNotFound){
  initCheckpoint
}
```

Reply:
fine with me